tools/functional-tester/etcd-tester/lease_stresser.go

```diff
@@ -0,0 +1,356 @@
+	if err != nil {
+		return fmt.Errorf("%v (%s)", err, ls.endpoint)
+	}
+	ls.conn = conn
+	ls.kvc = pb.NewKVClient(conn)
+	ls.lc = pb.NewLeaseClient(conn)
+
+	ls.aliveLeases = make(map[int64]time.Time)
+	ls.revokedLeases = make(map[int64]time.Time)
+	return nil
+}
+
+func (ls *leaseStresser) Stress() error {
+	plog.Infof("lease Stresser %v starting ...", ls.endpoint)
+	if err := ls.setupOnce(); err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	ls.cancel = cancel
+	ls.ctx = ctx
+
+	ls.runWg.Add(1)
+	go ls.run()
+	return nil
+}
+
+func (ls *leaseStresser) run() {
+	defer ls.runWg.Done()
+	ls.restartKeepAlives()
+	for ls.ctx.Err() == nil {
+		plog.Debugf("creating lease on %v ", ls.endpoint)
+		ls.createLeases()
+		plog.Debugf("done creating lease on %v ", ls.endpoint)
+		plog.Debugf("dropping lease on %v ", ls.endpoint)
+		ls.randomlyDropLeases()
+		plog.Debugf("done dropping lease on %v ", ls.endpoint)
+	}
+}
+
+func (ls *leaseStresser) restartKeepAlives() {
+	for leaseID := range ls.aliveLeases {
+		ls.aliveWg.Add(1)
+		go func(id int64) {
+			ls.keepLeaseAlive(id)
+		}(leaseID)
+	}
+}
+
+func (ls *leaseStresser) createLeases() {
+	neededLeases := ls.numLeases - len(ls.aliveLeases)
+	var wg sync.WaitGroup
+	wg.Add(neededLeases)
+	for i := 0; i < neededLeases; i++ {
+		go func() {
+			defer wg.Done()
+			leaseID, err := ls.createLease()
+			if err != nil {
+				plog.Errorf("lease creation error: (%v)", err)
+				return
+			}
+			plog.Debugf("lease %v created ", leaseID)
+			// if attaching keys to the lease encountered an error, we don't add the lease to the aliveLeases map
+			// because invariant check on the lease will fail due to keys not found
+			if err := ls.attachKeysWithLease(leaseID); err != nil {
+				return
+			}
+			ls.atomicAdd(leaseID, time.Now(), ls.aliveLeases)
+			// keep track of all the keep lease alive go routines
+			ls.aliveWg.Add(1)
+			go ls.keepLeaseAlive(leaseID)
+		}()
+	}
+	wg.Wait()
+}
+
+func (ls *leaseStresser) atomicAdd(leaseID int64, t time.Time, leases map[int64]time.Time) {
+	ls.rwLock.Lock()
+	leases[leaseID] = t
+	ls.rwLock.Unlock()
+}
+
+func (ls *leaseStresser) atomicRead(leaseID int64, leases map[int64]time.Time) (rv time.Time) {
+	ls.rwLock.RLock()
+	rv = leases[leaseID]
+	ls.rwLock.RUnlock()
+	return rv
+}
+
+func (ls *leaseStresser) atomicRemove(leaseID int64, leases map[int64]time.Time) {
+	ls.rwLock.Lock()
+	delete(leases, leaseID)
+	ls.rwLock.Unlock()
+}
+
+func (ls *leaseStresser) randomlyDropLeases() {
+	var wg sync.WaitGroup
+	for l := range ls.aliveLeases {
+		wg.Add(1)
+		go func(leaseID int64) {
+			defer wg.Done()
+			dropped, err := ls.randomlyDropLease(leaseID)
+			// if randomlyDropLease encountered an error such as context is cancelled, remove the lease from aliveLeases
+			// since we don't really know whether the lease is dropped or not.
+			if err != nil {
+				ls.atomicRemove(leaseID, ls.aliveLeases)
+				return
+			}
+			if !dropped {
```

Comment:
if an error has occurred, we have to check that first because the boolean value would be meaningless.